src/chip/index.tsx

```diff
@@ -0,0 +1,86 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+import { RenderResult } from '@dojo/framework/core/interfaces';
+import theme from '@dojo/framework/core/middleware/theme';
+import * as css from '../theme/chip.m.css';
+import Icon from '../icon/index';
+import { Keys } from '../common/util';
+
+export interface ChipProperties {
+	/** Renders an icon, provided with the value of the checked property */
+	iconRenderer?(checked?: boolean): RenderResult;
+	/** The label to be displayed in the widget */
+	label: string;
+	/** Renders a close icon, ignored if `onClose` is not provided */
+	closeRenderer?(): RenderResult;
+	/** A callback when the close icon is clicked, if `closeRenderer is not provided a default X icon will be used */
+	onClose?(): void;
+	/** An optional callback for the the widget is clicked */
+	onClick?(): void;
+	/** Whether the widget is disabled, only affects the widget when `onClick` is provided */
+	disabled?: boolean;
+	/** Indicates whe "checked" state of the widget, will be passed to the iconRenderer */
+	checked?: boolean;
+}
+
+const factory = create({ theme }).properties<ChipProperties>();
+
+export default factory(function Chip({ properties, middleware: { theme } }) {
+	const {
+		iconRenderer,
+		label,
+		closeRenderer,
+		onClose,
+		onClick,
+		disabled,
+		checked
+	} = properties();
+	const themedCss = theme.classes(css);
+
+	const createKeydown = (callback?: () => void) => (event: KeyboardEvent) => {
+		if (callback && (event.which === Keys.Enter || event.which === Keys.Space)) {
+			event.preventDefault();
+			callback();
+		}
+	};
+
+	const onCloseKeydown = createKeydown(onClose);
+
+	const clickable = !disabled && onClick;
+	return (
+		<div
+			key="root"
+			classes={[
+				themedCss.root,
+				disabled && themedCss.disabled,
+				clickable && themedCss.clickable
+			]}
+			role={clickable ? 'button' : undefined}
+			onclick={clickable ? onClick : undefined}
+			tabIndex={clickable ? 0 : undefined}
+			onkeydown={clickable ? createKeydown(onClick) : undefined}
+		>
+			{iconRenderer && iconRenderer(checked)}
+			<span classes={themedCss.label}>{label}</span>
+			{onClose && (
+				<span
+					key="closeButton"
+					classes={themedCss.closeIcon}
+					tabIndex={0}
+					role="button"
+					onclick={(event) => {
+						event.stopPropagation();
+						onClose();
+					}}
+					onkeydown={(event) => {
+						if (event.which === Keys.Enter || event.which === Keys.Space) {
+							event.stopPropagation();
+							onCloseKeydown(event);
```

Comment:
I'm not sure that we really need to expose the `key down` property for this widget. Also, would you not consider `Esc` to be a valid `close` key?

Reply:
If you tab over the close "button" and hit enter you'd expect it to close right? As for escape, I guess I'd never think to focus a button and click escape to close it but if that's expected behavior I can certainly add it. Should I handle that in the close button handler specifically or would handling escape anywhere on the chip component be more appropriate?

Reply:
I agree that key press should close it, just saying that I don't think we need to expose an `onKey` property as parent widgets etc won't need it.

Reply:
Just to clarify, we're not exposing any type of `onKey` property. The `onCloseKeydown` _was_ only internal and has since been removed.  It's just delegating to `onClick` when for Enter and Space keys.